File: joonas/078.cpp
#include <string>
#include <vector>
#include <cstring>

using namespace std;

const int INF = 0x3f3f3f3f;

int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    unsigned int d[201][201];
    memset(d, 0x3f, sizeof(d));
    for (int i = 1; i <= n; ++i) d[i][i] = 0;

    for (auto& e : fares) {
        d[e[0]][e[1]] = d[e[1]][e[0]] = e[2];
    }

    for (int k = 1; k <= n; ++k) {
        for (int i = 1; i <= n; ++i) {
            if (d[i][k] == INF) continue;
            for (int j = 1; j <= n; ++j) {
                d[i][j] = min(d[i][j], d[i][k] + d[k][j]);
            }
        }
    }

    unsigned int ans = INF;
    for (int i = 1; i <= n; ++i) {
        ans = min(ans, d[s][i] + d[i][a] + d[i][b]);
    }
    return ans;
}

File: jlogkim/Season1/078.java
class Solution {
    public int solution(int n, int s, int a, int b, int[][] fares) {
        long dists[][] = new long[n][n];
        for(int i=0;i<n;i++) {
        	for(int j=0;j<n;j++)if(i!=j)dists[i][j]=Integer.MAX_VALUE;
        }
        
        for(int i=0;i<fares.length;i++) {
        	dists[fares[i][0]-1][fares[i][1]-1]=fares[i][2];
        	dists[fares[i][1]-1][fares[i][0]-1]=fares[i][2];
        }
        for(int i=0;i<n;i++) {
        	for(int st=0;st<n;st++) {
        		for(int ed=0;ed<n;ed++) {
        			dists[st][ed]=Math.min(dists[st][ed],dists[st][i]+dists[i][ed]);
        		}
        	}
        }
        
        long ans=dists[s-1][a-1]+dists[s-1][b-1];
        long tmp;
        for(int i=0;i<n;i++) {
        	tmp=dists[s-1][i]+dists[i][a-1]+dists[i][b-1];
        	if(tmp<ans)ans=tmp;
        }
        
        
        return (int)ans;
    }
}